File: sources/Android/library/src/main/java/com/lua/mobile/sdk/library/module/TasksModule.java
/**
 * Android LuaMobileSdk for Android framework project.
 *
 * Copyright 2016 Arno Zhang <zyfgood12@163.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lua.mobile.sdk.library.module;

import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;
import android.util.SparseArray;
import com.lua.mobile.sdk.engine.bridge.LuaMethod;
import com.lua.mobile.sdk.engine.bridge.LuaModuleBridge;
import com.lua.mobile.sdk.engine.lua.LuaObject;

public class TasksModule extends LuaModuleBridge {

    public static final int invalid_id = -1;


    private Handler mHandler = new Handler(Looper.getMainLooper());
    private SparseArray<Runnable> mTimerRunnable = new SparseArray<>();
    private int mCurrentIndex = 0;


    @Override
    public String getBridgeName() {
        return "Tasks";
    }

    @LuaMethod
    public void async(final Object object) {
        if (object instanceof LuaObject) {
            @SuppressLint("StaticFieldLeak")
            AsyncTask<Void, Void, Void> task = new AsyncTask<Void, Void, Void>() {
                @Override
                protected Void doInBackground(Void... params) {
                    LuaObject callback = (LuaObject) object;
                    callback.call();
                    return null;
                }
            };

            task.execute();
        }
    }

    @LuaMethod
    public void post(Object object) {
        postTo(new Handler(Looper.myLooper()), object);
    }

    @LuaMethod
    public void postToMain(Object object) {
        postTo(mHandler, object);
    }

    @LuaMethod
    public boolean isMainThread() {
        return Looper.myLooper() == Looper.getMainLooper();
    }

    @LuaMethod
    public int setTimer(
            final Object object, final int millSeconds, final boolean single, int delay) {

        if (object instanceof LuaObject) {
            final int timerId = nextTimerId();
            Runnable runnable = new Runnable() {
                @Override
                public void run() {
                    if (single) {
                        removeTimer(timerId);
                    } else {
                        mHandler.postDelayed(this, millSeconds);
                    }

                    LuaObject callback = (LuaObject) object;
                    callback.call(timerId);
                }
            };

            mTimerRunnable.put(timerId, runnable);
            mHandler.postDelayed(runnable, millSeconds);
            return timerId;
        }

        return invalid_id;
    }

    @LuaMethod
    public void cancelTimer(int timerId) {
        Runnable runnable = mTimerRunnable.get(timerId);
        removeTimer(timerId);
        if (runnable != null) {
            mHandler.removeCallbacks(runnable);
        }
    }

    @Override
    public void destroy() {
        super.destroy();
        mHandler.removeCallbacksAndMessages(null);
    }

    private static void postTo(Handler handler, final Object object) {
        if (object instanceof LuaObject) {
            handler.post(new Runnable() {
                @Override
                public void run() {
                    LuaObject callback = (LuaObject) object;
                    callback.call();
                }
            });
        }
    }

    private synchronized int nextTimerId() {
        return ++mCurrentIndex;
    }

    private synchronized void removeTimer(int timerId) {
        mTimerRunnable.remove(timerId);
    }
}

File: sources/iOS/LuaMobileSdk/LuaMobileSdk/Sources/Library/module/TasksModule.m
/**
 * Android LuaMobileSdk for iOS framework project.
 *
 * Copyright 2017 Arno Zhang <zyfgood12@163.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "TasksModule.h"
#import "LuaObject.h"
#import "BridgeMethodDefine.h"
#import "CommonUtils.h"
#import "BoolObject.h"


static const int invalid_id = -1;


#define SDK_BRIDGE_IMPL TasksModule

SDK_STATIC_METHOD_IMPL_BEGIN()

SDK_STATIC_RET_0_PARAM_1(async, toObjectFixed)
SDK_STATIC_RET_0_PARAM_1(post, toObjectFixed)
SDK_STATIC_RET_0_PARAM_1(postToMain, toObjectFixed)
SDK_STATIC_RET_1_PARAM_0(isMainThread)
SDK_STATIC_RET_0_PARAM_1(cancelTimer, toIntFixed)

SDK_STATIC_METHOD_IMPL_END()


NSObject* _TasksModule_setTimer(LuaMethodInvokeEnv* env, LuaBridgeArgs* args) {
    TasksModule* module = (TasksModule*) env.bridge;
    int timerId = [module setTimer:[args toObjectFixed:1] Interval:[args toIntFixed:2]
                            Single:[args toBool:3] Delay:[args toIntFixed:4]];
    return @(timerId);
}


@implementation TasksModule {
    int mCurrentIndex;
    NSMutableDictionary<NSNumber*, NSTimer*>* mTimers;
    dispatch_queue_t mAsyncQueue;
}


- (instancetype)init {
    if (self = [super init]) {
        mTimers = [[NSMutableDictionary alloc] init];
        mAsyncQueue = dispatch_queue_create("LuaMobileSdk.Tasks", NULL);
    }

    return self;
}

- (NSString*)getBridgeName {
    return @"Tasks";
}

- (void)registerStaticVariables:(id <StaticVariablesRegHelper>)helper {
    [super registerStaticVariables:helper];

    SDK_STATIC_VARIABLE_INT(invalid_id);
}

- (void)registerStaticMethods:(id <MethodsRegHelper>)helper {
    [super registerStaticMethods:helper];

    SDK_STATIC_METHOD(post);
    SDK_STATIC_METHOD(async);
    SDK_STATIC_METHOD(setTimer);
    SDK_STATIC_METHOD(cancelTimer);
    SDK_STATIC_METHOD(postToMain);
    SDK_STATIC_METHOD(isMainThread);
}

- (void)async:(NSObject*)callback {
    if (!mAsyncQueue || ![callback isKindOfClass:[LuaObject class]]) {
        return;
    }

    dispatch_async(mAsyncQueue, ^() {
        LuaObject* cbk = (LuaObject*) callback;
        [cbk call:nil];
    });
}

- (void)post:(NSObject*)callback {
    if (![callback isKindOfClass:[LuaObject class]]) {
        return;
    }

    [CommonUtils dispatchToCurrentQueue:^() {
        LuaObject* cbk = (LuaObject*) callback;
        [cbk call:nil];
    }];
}

- (void)postToMain:(NSObject*)callback {
    if (![callback isKindOfClass:[LuaObject class]]) {
        return;
    }

    [CommonUtils dispatchToMain:^{
        LuaObject* cbk = (LuaObject*) callback;
        [cbk call:nil];
    }];
}

- (BoolObject*)isMainThread {
    return [BoolObject valueOf:[CommonUtils isMainThread]];
}

- (int)setTimer:(NSObject*)callback
       Interval:(int)intervalMillSeconds
         Single:(BOOL)single
          Delay:(int)delayMillSeconds {

    @synchronized (self) {
        return [self setTimerInternal:callback Interval:intervalMillSeconds
                               Single:single Delay:delayMillSeconds];
    }
}

- (int)setTimerInternal:(NSObject*)callback
               Interval:(int)intervalMillSeconds
                 Single:(BOOL)single
                  Delay:(int)delayMillSeconds {

    if (![callback isKindOfClass:[LuaObject class]]) {
        return invalid_id;
    }

    int timerId = [self nextTimerId];
    void (^block)(NSTimer* timer) = ^(NSTimer* timer_) {
        if (single) {
            [timer_ invalidate];
            [mTimers removeObjectForKey:@(timerId)];
        }

        LuaObject* cbk = (LuaObject*) callback;
        [cbk call:@(timerId), nil];
    };

    NSTimer* timer = nil;
    double interval = intervalMillSeconds / 1000.0;
    if (delayMillSeconds > 0) {
        NSDate* date = [[NSDate alloc] initWithTimeIntervalSinceNow:delayMillSeconds / 1000.0];
        timer = [[NSTimer alloc] initWithFireDate:date interval:interval repeats:!single block:block];
    } else {
        timer = [NSTimer scheduledTimerWithTimeInterval:interval repeats:!single block:block];
    }

    mTimers[@(timerId)] = timer;
    [[NSRunLoop mainRunLoop] addTimer:timer forMode:NSRunLoopCommonModes];
    return timerId;
}

- (void)cancelTimer:(int)timerId {
    @synchronized (self) {
        NSTimer* timer = mTimers[@(timerId)];
        if (!timer) {
            return;
        }

        [timer invalidate];
        [mTimers removeObjectForKey:@(timerId)];
    }
}

- (int)nextTimerId {
    return ++mCurrentIndex;
}

- (void)destroy {
    [super destroy];

    mAsyncQueue = nil;

    NSEnumerator<NSTimer*>* timers = [mTimers objectEnumerator];
    for (NSTimer* timer in timers) {
        [timer invalidate];
    }

    [mTimers removeAllObjects];
}

@end